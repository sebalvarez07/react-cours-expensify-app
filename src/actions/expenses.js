import moment from 'moment';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        // Setup defaults. If no description than empty string etc.
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = moment(0).valueOf() 
        } = expenseData;

        const expense = {description, note, amount,createdAt };

        // Push().then() comes back with a reference (red) to the ojbect that was just added (in this case the id)
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
};  

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
        type: 'EDIT_EXPENSE',
        id, 
        updates
});

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});