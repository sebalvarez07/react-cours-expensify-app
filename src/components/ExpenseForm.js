import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString(): '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    }

    onDescriptionChange = (e) => {

        /* 
            We do this as a two step process because setState is an async function meaning that the object contents do not get executed right away
            This is a problem in this case as the value of e.target does not 'persist', and must be used right away.
            There is another way to go about this if we want to use description: e.target.value inside the setState
            By putting this line before everything :
            e.persist()
        */
       
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        // The $ sign at the end is crucial as it marks the end of the input. Anything written after the 
        // 2 decimal places will not match and so it will not be written on the state, which means it will
        // Not be printed as a value to the input
        // Thus the input is registered onChange, but not displayed, the process of dealing with the input
        // Handled and the when the state is updated, the value={this.state.amount} will be printed

        /*
            ^   - marks the begining of input
            \d  - Means digits = [0 - 9]
            *   - Means match the preceeding expression 0 or more times = {0,}
            ()? - Means match preceeding expression 0 or 1 time (maybe it happens, maybe not) = {0, 1}
            \.  - Means literally a dot(.)
            $   - Marks the end of input. IF this isn't the end, returns false

            !amount --> If there is no amount OR if there is a match
        */

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({ amount }));
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onDateChange = (createdAt) => {
        
        if(createdAt){
            this.setState(() => ({ createdAt: createdAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    handleSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error: 'Please provide a description and amount'}));
        } else {
            this.setState(() => ({error: '' }));
            this.props.onSubmit({
                description: this.state.description, 
                amount: parseFloat(this.state.amount, 10) * 100,    // We eliminate the . of decimals while keeping the decimes in the ciphre
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>

                {this.state.error && <h3>{this.state.error}</h3>}

                <label>Description</label>
                <input 
                    type="text" 
                    name="description" 
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
        
                <label>Amount</label>
                <input 
                    type="text" 
                    name="amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
        
                <label>Note</label>
                <textarea 
                    name="note"
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                />

                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange} 
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
        
                <button>Add Expense</button>
            </form>
        )
    }
}

export default ExpenseForm;