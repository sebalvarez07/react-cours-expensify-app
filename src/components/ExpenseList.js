import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenes from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map(
                        (expense, index) => <ExpenseListItem key={`${expense.id}`} {...expense}/>
                    )
                )
            }
        </div>
        
    </div> 
);

/*
    - This is called everytime the store changes
    - Returns an object with the data that this component needs and sends it to the component as props
    - Returns an object with expenses and dispatch fn
*/
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenes(state.expenses, state.filters)
    };
};
/*
    - connect() returns a new function that takes the component to be wrapped as an argument
    - We don't have to be concerned with store.subscribe() or dispatching. All taken care of by react-redux
*/
export default connect(mapStateToProps)(ExpenseList);