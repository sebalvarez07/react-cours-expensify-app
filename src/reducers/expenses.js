// Expenses Reducer

const expensesReduceDefault = [];

// The state is simply an array --> the expense object is created inside the combineReducers() call
const expensesReducer = (state = expensesReduceDefault, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            // On the filter parameters we are destructuring currentElement.id, by doing { id } = 
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                // console.log(action.expense.id);
                    if(expense.id === action.id) {
                        return {
                            ...expense,
                            ...action.updates
                        }
                    } else {
                        return expense;
                    }
                })
            
        default: 
            return state
    }
}

export default expensesReducer;