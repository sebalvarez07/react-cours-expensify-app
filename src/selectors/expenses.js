// Get visible expenses
import moment from 'moment';

const getVisibleExpenes = (expenses, {text, sortBy, startDate, endDate}) => {

    const temp =  expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        // Statement --> if startDate undefined OR expense was created a date later than filter start date
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true ;
        // If text filter is empty it will come back as true anyway '' is not the same as ' '
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    });

    temp.sort( (a, b) => {
        if(sortBy === 'date'){  
            // return a.createdAt - b.createdAt
            return a.createdAt < b.createdAt? 1 : -1;
        }

        else if(sortBy === 'amount'){
            // return a.amount - b.amount
            return a.amount < b.amount ? 1 : -1;
        }
    });

    return temp;
}

export default getVisibleExpenes;