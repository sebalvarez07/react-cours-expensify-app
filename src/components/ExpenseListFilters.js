import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    handleOnDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocuseChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    };

    onChangeTextFilter = (e) => {
        const val = e.target.value;
        this.props.setFilterText(val);
    };

    onChangeSortbyFilter = (e) => {
        const val = e.target.value;

        if(val === 'date') {
            this.props.sortByDate();
        } 
        else if(val === 'amount') {
            this.props.sortByAmount();
        }
    };

    render(){
        return (
            <div>
                <input 
                    type="text"  
                    value={this.props.filters.text} 
                    onChange={this.onChangeTextFilter}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onChangeSortbyFilter}>
                    
                    <option value="amount">Amount</option>
                    <option value="date">Date</option>
                </select>
        
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate} 
                    onDatesChange={this.handleOnDatesChange}
                    focusedInput={this.state.calendarFocused} 
                    onFocusChange={this.onFocuseChange}
                    isOutsideRange={() => false}
                    numberOfMonths={1}
                    showClearDates={true}
                />
            </div>
        );
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        setFilterText: (text) => dispatch(setFilterText(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const ExpenseListFiltersWrapped = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ExpenseListFiltersWrapped;