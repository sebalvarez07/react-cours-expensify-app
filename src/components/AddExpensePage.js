import React from 'react';
import ExpenseForm from './ExpenseForm';
import  { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        {/** 
            We use redirect, as we do not have a Link that we can simply click to send us to a new page
            So we use history.push (which we get as a prop for being a Router component - see AppRouter.js under the routers folder)
            This redirects us to any of the paths that are Router components

            NOTE: Using push also allow us to go back through the browsers back button as if it were actually a multi-page site
        */}
        this.props.history.push('/');           
    }

    render() {
        return (
            <div>  
                <h2>This is from my add Expense component</h2>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        )
    }
}

// Function gives us access to dispatch, and allows us to return an object that turns its properties into props properties
const mapDispatchToProps = (dispatch) => ({ startAddExpense: (expense) => dispatch(startAddExpense(expense)) });

export default connect(undefined, mapDispatchToProps)(AddExpensePage);