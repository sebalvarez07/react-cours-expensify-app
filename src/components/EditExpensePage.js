import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
 
export class EditExpensePage extends React.Component {
    
    onSubmit = (updates) => {
        this.props.startEditExpense(this.props.expense.id, updates);
        this.props.history.push('/');
    };

    handleRemoveExpense = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push('/');
    };

    render() {
        return (
            <div>

                <div class="page-header">
                    <div class="content-container">
                        <h2 class="page-header__title">
                            Edit Expense
                        </h2>
                    </div> 
                </div>

                <div class="content-container">
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="btn btn--grey" onClick={this.handleRemoveExpense}>
                        Remove Expense
                    </button>
                </div>
            </div>
        )
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates)),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    }
}

/*
    state gives the store's state
    props gives us the connected (or wrapped) props
*/
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(el => el.id === props.match.params.id)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);