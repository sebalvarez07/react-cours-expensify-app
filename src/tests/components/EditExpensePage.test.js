import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let startEditExpense, history, wrapper, startRemoveExpense;

beforeEach( () => {
    startRemoveExpense = jest.fn();
    startEditExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage 
        expense={expenses[0]} 
        startEditExpense={startEditExpense} 
        history={history} 
        startRemoveExpense={startRemoveExpense}/>);
});

test('Should render edit expense', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle edit expense on submit', () => {

    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('Should handle remove on button clicked', () => {
    wrapper.find('.remove-expense__btn').simulate('click');

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[0].id});
})