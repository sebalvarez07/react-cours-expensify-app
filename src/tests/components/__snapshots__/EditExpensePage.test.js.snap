// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense 1`] = `
<div>
  <div
    class="page-header"
  >
    <div
      class="content-container"
    >
      <h2
        class="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    class="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="btn btn--grey"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
